### PR TITLE
Introduce pylint to pynessie in order to enforce more strict checking

### PR DESCRIPTION
**Pull request by omarsmak**
This PR intends to introduce `pylint` in order to enforce more strict static typing, this can work alongside `flake8` and `mypy`

### CODE REVIEW

File: python/.pylintrc [hunk 1]
```diff
@@ -0,0 +1,574 @@
+[MASTER]
```
Comment (snazy) [inline python/.pylintrc:1]: Question: where does this config file come from? (I'm in no way a Python guru - so bear with me ;) )

File: python/tox.ini [hunk 2]
```diff
@@ -35,6 +35,15 @@ commands =
     safety check --ignore=41002
     mypy --ignore-missing-imports -p pynessie
 
+[testenv:pylint]
+basepython = python
+deps =
+    -r{toxinidir}/requirements_dev.txt
```
[the remark this answers is not in the document]
Comment (omarsmak) [reply python/tox.ini:41]: It should be enough, in fact only `requirements_lint.txt` is needed, I will change this

File: python/.pylintrc (hunk 1 above)
Comment (omarsmak) [reply python/.pylintrc:1]: no problems :). It is generated through `pylint` cli via this command `pylint --generate-rcfile > .pylintrc` and then from there, the config file can be modified as needed

Comment (snazy) [reply python/.pylintrc:1]: So it's the default for now, right?

**Comment by snazy** [comment]
I'm okay with this change, having stricter type checks and such things is a good thing IMO.
Maybe just mention it to @rymurr & @dimas-b whether they're okay with it as well.

File: python/.pylintrc (hunk 1 above)
Comment (omarsmak) [reply python/.pylintrc:1]: Nope, I have disabled some features like `consider-using-f-string`, `too-few-public-methods`  .. etc as I think they are not necessary needed for now and changed properties like `max-line-length`

File: python/tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (omarsmak) [reply python/tox.ini]: fixed

**Review by snazy — APPROVED** [review]

File: python/pynessie/error.py [hunk 3]
```diff
@@ -168,22 +168,22 @@ def _create_nessie_exception(error: dict, status: int, reason: str, url: str) ->
...
-    elif "REFERENCE_CONFLICT" == error_code:
+    if error_code == "REFERENCE_CONFLICT":
         return NessieReferenceConflictException(error, status, url, reason)
     return None
 
 
 def _create_exception(error: dict, status: int, reason: str, url: str) -> Exception:
     if 400 <= status < 500:
-        reason = u"Client Error (%s)" % reason
+        reason = f"Client Error {reason}"
```
Comment (dimas-b) [inline python/pynessie/error.py:184]: TIL :) This looks a lot nicer :+1:

File: python/tox.ini [hunk 4]
```diff
@@ -15,33 +15,37 @@
...
 [tox]
-envlist = py36, py37, py38, py39, flake8
+envlist = py36, py37, py38, py39, flake8, pylint
 
 [gh-actions]
 python =
     3.6: py36
     3.7: py37
-    3.8: py38, flake8
+    3.8: py38, flake8, pylint
```
Comment (dimas-b) [inline python/tox.ini:24]: just wondering: why do we run style tests on 3.8 and not 3.9? I'd guess going with the latest version for style-related stuff would be preferable from the general code evolution perspective?

File: python/tox.ini [hunk 5]
```diff
@@ -15,33 +15,37 @@
...
     flake8 pynessie tests
     safety check --ignore=41002
     mypy --ignore-missing-imports -p pynessie
 
+[testenv:pylint]
+basepython = python
+deps =
+    -r{toxinidir}/requirements_lint.txt
+commands =
+    pylint pynessie
```
Comment (dimas-b) [inline python/tox.ini:41]: why not lint tests too (in the spirit of "test is code")?

**Review by dimas-b — COMMENTED** [review]
LGTM overall :100: Just wondering if we should `pylint` the test code too. Plus a naming comment about `y_s` and related variables.

File: python/tox.ini (hunk 4 above)
Comment (omarsmak) [reply python/tox.ini:24]: good question, from my side, I have no idea 😄.

File: python/tox.ini (hunk 5 above)
Comment (omarsmak) [reply python/tox.ini:41]: good catch, will add that as well

File: python/pynessie/commands/branch.py [hunk 6]
```diff
@@ -25,8 +25,8 @@
 
 
 @click.command(name="branch")
-@click.option("-l", "--list", cls=MutuallyExclusiveOption, is_flag=True, help="list branches", mutually_exclusive=["delete"])
-@click.option("-d", "--delete", cls=MutuallyExclusiveOption, is_flag=True, help="delete a branch", mutually_exclusive=["list"])
+@click.option("-l", "--list", "is_list", cls=MutuallyExclusiveOption, is_flag=True, help="list branches", mutually_exclusive=["delete"])
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply python/pynessie/commands/branch.py:28]: fair enough

File: python/pynessie/client/nessie_client.py [hunk 7]
```diff
@@ -228,7 +228,7 @@ def get_log(
         """
         page_token = filtering_args.get("pageToken", None)
 
-        def fetch_logs(max: Optional[int], token: Optional[str] = page_token) -> LogResponse:
+        def fetch_logs(max_: Optional[int], token: Optional[str] = page_token) -> LogResponse:
```
[the remark this answers is not in the document]
Comment (omarsmak) [reply python/pynessie/client/nessie_client.py:231]: Fixed

File: python/pynessie/conf/__init__.py [hunk 8]
```diff
@@ -29,7 +29,7 @@
 
 def to_dict(config: confuse.Configuration) -> dict:
     """Convert confuse Configuration object to dict."""
-    ys = config.dump(redact=True)
-    fd = StringIO(ys)
-    dct = yaml.safe_load(fd)
+    y_s = config.dump(redact=True)
+    f_d = StringIO(y_s)
```
[the remark this answers is not in the document]
Comment (omarsmak) [reply python/pynessie/conf/__init__.py:33]: Changed it to something more meaningful

File: python/tox.ini (hunk 5 above)
Comment (omarsmak) [reply python/tox.ini]: @dimas-b I have added pylint to the tests. However, I had to exchange the global variables with dataclass as pylint wasn't happy about using the global variables. Can you please check this out?

File: python/tests/conftest.py [hunk 9]
```diff
@@ -98,15 +103,15 @@ def reset_nessie_server_state() -> None:
...
+    if nessie_test_config.cleanup:
         return None
     return request
 
 
 @pytest.fixture(scope="module")
 def vcr_config() -> dict:
     """VCR config that adds a custom before_record_request callback."""
-    pytest.nessie_cleanup = False
+    nessie_test_config.cleanup = False
```
Comment (dimas-b) [inline python/tests/conftest.py:114]: nice catch :+1:

**Review by dimas-b — APPROVED** [review]

**Review by nastra — APPROVED** [review]
overall LGTM and I like the idea of doing more strict checking

**Pull request merged**